### PR TITLE
Optionally run random tick in cubes

### PR DESCRIPTION
**Pull request by Foghrye4**

### CODE REVIEW

File: src/main/java/cubicchunks/world/cube/Cube.java [hunk 1]
```diff
@@ -375,7 +382,7 @@ public void tickCubeCommon(BooleanSupplier tryToTickFaster) {
      * @param rand - World specific Random
      */
     public void tickCubeServer(BooleanSupplier tryToTickFaster, Random rand) {
-        if (!isFullyPopulated) {
+        if (!isFullyPopulated || this.storage == null) {
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/world/cube/Cube.java]: If storage is null light may still need updates

File: src/main/java/cubicchunks/asm/mixin/selectable/common/MixinWorldServer_UpdateBlocks.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import net.minecraft.world.chunk.Chunk;
+
+@Mixin(value = WorldServer.class, priority = 1001)
+public abstract class MixinWorldServer_UpdateBlocks implements ICubicWorldServer {
+
+    @Shadow
+    protected abstract void playerCheckLight();
+
+    @Inject(method = "updateBlocks", at = @At("HEAD"), cancellable = true)
+    public void onUpdateBlocks(CallbackInfo ci) {
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/mixin/selectable/common/MixinWorldServer_UpdateBlocks.java]: Could you instead attempt to redirect the right methods to no-ops here or using ModifyVariable? This overwrite would need updating whenever this method is changed and it's a big method. And it's likely that mods will attempt to modify it.

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java [hunk 3]
```diff
@@ -169,6 +170,11 @@ public boolean testForCubes(CubePos start, CubePos end, Predicate<Cube> cubeAllo
         return this.chunkProvider.provideChunk(blockToCube(blockX), blockToCube(blockZ))
                 .getHeightValue(blockToLocal(blockX), blockToLocal(blockZ));
     }
+    
+    @Override public int updateLCG() {
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java]: Why this method is there? It's only used in World mixins.

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java (hunk 3 above)
Comment (Foghrye4) [reply src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java]: If I will redirect only random tick, it will not be necessary.

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java]: Isn't there a vanilla method that does the same thing? You could use `@Shadow` is there is. If there isn't, you are adding unnecessary changes to `updateBlocks`. Either way it shouldn't be there.

**Comment by Foghrye4** [comment]
Done as requested.

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 4]
```diff
@@ -567,7 +573,7 @@ public void flush() throws IOException {
     }
 
     @SuppressWarnings("unchecked")
-    Iterator<IColumn> columnsIterator() {
+    public Iterator<IColumn> columnsIterator() {
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: Why?

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 5]
```diff
@@ -190,11 +190,17 @@ public boolean saveChunks(boolean alwaysTrue) {
     @Override
     public boolean tick() {
         // NOTE: the return value is completely ignored
-        profiler.startSection("providerTick("+cubeMap.getSize()+")");
+        profiler.startSection("providerTick(" + cubeMap.getSize() + ")");
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: You should have done it right when writing this for the first time

File: src/main/java/cubicchunks/asm/mixin/selectable/common/MixinWorldServer_UpdateBlocks.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import cubicchunks.world.ICubicWorldServer;
+import net.minecraft.world.GameRules;
+import net.minecraft.world.WorldServer;
+
+@Mixin(value = WorldServer.class, priority = 1001)
+public abstract class MixinWorldServer_UpdateBlocks implements ICubicWorldServer {
+
+    @Redirect(method = "updateBlocks", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/GameRules;getInt(Ljava/lang/String;)I"), require = 1)
+    public int redirectGetRandomTickSpeed(GameRules gameRules, String ruleName) {
+        return this.isCubicWorld() ? 0 : gameRules.getInt(ruleName);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/mixin/selectable/common/MixinWorldServer_UpdateBlocks.java]: Add comment explaining where to look for the real code that does it

File: src/main/java/cubicchunks/asm/CubicChunksMixinConfig.java [hunk 7]
```diff
@@ -144,5 +153,46 @@ public static String getNicelyFormattedName(String name) {
...
+    
+    private void writeConfigToJson(File configFile) throws IOException {
+        JsonWriter writer = new JsonWriter(new FileWriter(configFile));
+        writer.setIndent(" ");
+        writer.beginArray();
+        for (BoolOptions configOption : BoolOptions.values()) {
+            writer.beginObject();
+            writer.name("name");
+            writer.value(configOption.name());
+            writer.name("description");
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/CubicChunksMixinConfig.java]: Gson should support comments is you enable that feature, could you use that instead and simply store the values in normal way?

something like

```java
/* create comment in whatever way needed */
writer.name(configOption.name());
writer.value(configOption.value);
```

**Review by Niko-sk2x — COMMENTED** [review]
Other than that 'm fine with the code

File: src/main/java/cubicchunks/asm/CubicChunksMixinConfig.java (hunk 7 above)
Comment (Foghrye4) [reply src/main/java/cubicchunks/asm/CubicChunksMixinConfig.java]: I did not find any mention on how to add a comments of `\\` or `\**\` type in Json files. Even more, I found an info that such comments are not allowed by Json syntax. Existing specifications of JSON format state that JSON must contain tokens and only tokens.
`writer.name("description");` or any other token ignored by parser is acceptable way of adding comments in JSON files.

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/asm/CubicChunksMixinConfig.java]: Looks like Gson supports comments, but there doesn't seem to be a way to write them using the writer. I knowthat json specification doesn't allow them. Then leave the description the way it is, but name and value shoudl be one thing, it makes no sense to do what you did here.

**Comment by Foghrye4** [comment]
Generated JSON:
```
[
 {
  "USE_FAST_COLLISION_CHECK": true,
  "description": "Enabling this option allow using fast collision check. Fast collision check can reduce server lag. You need to restart Minecraft to apply changes."
 },
 {
  "RANDOM_TICK_IN_CUBE": true,
  "description": "If set to true, random tick wil be launched from cube instance instead of chunk. Cube based random tick may slightly reduce server lag. You need to restart Minecraft to apply changes."
 }
]
```

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 8]
```diff
@@ -192,9 +192,15 @@ public boolean tick() {
...
         long i = System.currentTimeMillis();
+        int randomTickSpeed = this.world.getGameRules().getInt("randomTickSpeed");
         Random rand = this.world.rand;
         for (Cube cube : cubeMap) {
             cube.tickCubeServer(() -> System.currentTimeMillis() - i > 40, rand);
+            if (cube.isEmpty())
+                continue;
+            int randomTickCounter = randomTickSpeed;
+            while (randomTickCounter-- > 0)
+                cube.randomTick(this.world, rand);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: I don't know how I missed it before, but won't it also run when the config option says it should use vanilla code?

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]

File: src/main/java/cubicchunks/server/CubeProviderServer.java (hunk 8 above)
Comment (Foghrye4) [reply src/main/java/cubicchunks/server/CubeProviderServer.java]: Yes.

Comment (Foghrye4) [reply src/main/java/cubicchunks/server/CubeProviderServer.java]: @Barteks2x You mean it should not? Why not?

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/server/CubeProviderServer.java]: Because if both run then blocks will be ticked twice per tick

**Comment by Foghrye4** [comment]
It will be tick in different blocks. Because we using different random seeds in cube random tick and world random tick. But whatever. I changed it.

**Comment by Niko-sk2x** [comment]
But it would mean that for example crops would grow on average twice as fast. So it had to be  fixed.

**Review by Niko-sk2x — APPROVED** [review]

**Pull request merged**